### ISSUE TITLE
Fresh install of ludwig does not work without torchtext and torchvision.  Also psutil missing from requirements.txt

### ISSUE THREAD

**Issue by dantreiman** [ludwig-ai/ludwig#1914]
TL;DR: I recommend we move `torchvision`, `torchtext` and `ray` into the default requirements.txt

My reasoning is this: new users should be able to use all of the features of Ludwig without extra steps, including text, images, and ray parallelism.

---

Ludwig's code today does not isolate the text and vision dependencies, and this easily goes unnoticed because most DL environments with torch have torchtext and torchvision installed already.  Though we could untangle and encapsulate these python deps I do not think it is worth the added complexity.

I think the cost of installing `torchvision` and `torchtext` is trivial compared to the engineering cost of maintaining optional dependencies on these packages and gracefully handling errors when these packages are not installed.

If someone is developing a container, and they aren't using images and they really need to forgo installing `torchvision` to save space, that is a 1% case, and we can provide a way to do that but it should not be the default.  Maybe we could make a `ludwig-serve` package that strips away all but the minimial dependencies for serving?  Just a thought.

---

Here is the log of a new install I just ran today:

In a new conda environment `temp39`, I installed ludwig `pip install -e .` then try to train higgs:

```
(temp39) daniel@Daniel-M1-Pro higgs % python train_higgs_small.py
Note: NumExpr detected 10 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
NumExpr defaulting to 8 threads.
Traceback (most recent call last):
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/examples/tabnet/higgs/train_higgs_small.py", line 3, in <module>
    from ludwig.api import LudwigModel
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/api.py", line 38, in <module>
    from ludwig.backend import Backend, initialize_backend
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/backend/__init__.py", line 20, in <module>
    from ludwig.backend.base import Backend, LocalBackend
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/backend/base.py", line 25, in <module>
    from ludwig.models.ecd import ECD
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/models/ecd.py", line 10, in <module>
    from ludwig.combiners.combiners import Combiner, get_combiner_class
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/combiners/combiners.py", line 25, in <module>
    import ludwig.marshmallow.marshmallow_schema_utils as schema
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/marshmallow/marshmallow_schema_utils.py", line 10, in <module>
    from ludwig.modules.reduction_modules import reduce_mode_registry
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/modules/reduction_modules.py", line 19, in <module>
    from ludwig.modules.attention_modules import FeedForwardAttentionReducer
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/modules/attention_modules.py", line 21, in <module>
    from ludwig.utils.torch_utils import get_activation, LudwigModule
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/utils/torch_utils.py", line 12, in <module>
    from ludwig.utils.strings_utils import SpecialSymbol
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/utils/strings_utils.py", line 29, in <module>
    from ludwig.utils.tokenizers import tokenizer_registry
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/utils/tokenizers.py", line 21, in <module>
    import torchtext
ModuleNotFoundError: No module named 'torchtext'
```

To fix this, I install the text dependencies:  `pip install -e ".[text]"`

After installing `torchtext`:

```
(temp39) daniel@Daniel-M1-Pro higgs % python train_higgs_small.py
Note: NumExpr detected 10 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
NumExpr defaulting to 8 threads.
Traceback (most recent call last):
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/examples/tabnet/higgs/train_higgs_small.py", line 3, in <module>
    from ludwig.api import LudwigModel
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/api.py", line 38, in <module>
    from ludwig.backend import Backend, initialize_backend
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/backend/__init__.py", line 20, in <module>
    from ludwig.backend.base import Backend, LocalBackend
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/backend/base.py", line 25, in <module>
    from ludwig.models.ecd import ECD
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/models/ecd.py", line 10, in <module>
    from ludwig.combiners.combiners import Combiner, get_combiner_class
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/combiners/combiners.py", line 27, in <module>
    from ludwig.encoders.sequence_encoders import ParallelCNN, StackedCNN, StackedCNNRNN, StackedParallelCNN, StackedRNN
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/encoders/__init__.py", line 8, in <module>
    import ludwig.encoders.image_encoders
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/encoders/image_encoders.py", line 24, in <module>
    from ludwig.modules.convolutional_modules import Conv2DStack, ResNet
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/modules/convolutional_modules.py", line 22, in <module>
    from ludwig.utils.image_utils import get_img_output_shape
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/ludwig/utils/image_utils.py", line 26, in <module>
    import torchvision.transforms.functional as F
ModuleNotFoundError: No module named 'torchvision'
```

After installing `pip install -e ".[image]"`

```
(temp39) daniel@Daniel-M1-Pro higgs % python train_higgs_small.py
Note: NumExpr detected 10 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
NumExpr defaulting to 8 threads.
import ray failed with exception: No module named 'ray'
Traceback (most recent call last):
  File "/Users/daniel/Desktop/github/dantreiman-ludwig/examples/tabnet/higgs/train_higgs_small.py", line 3, in <module>
    from ludwig.api import LudwigModel
  File "/Users/daniel/mambaforge/envs/temp39/lib/python3.9/site-packages/ludwig/api.py", line 66, in <module>
    from ludwig.models.predictor import (
  File "/Users/daniel/mambaforge/envs/temp39/lib/python3.9/site-packages/ludwig/models/predictor.py", line 10, in <module>
    import psutil
ModuleNotFoundError: No module named 'psutil'
```

Oops, we also need to install `psutil`. `pip install psutil`:

Ok now it trains, but I don't have ray so it only uses one core:
```
(temp39) daniel@Daniel-M1-Pro higgs % python train_higgs_small.py
Note: NumExpr detected 10 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
NumExpr defaulting to 8 threads.
import ray failed with exception: No module named 'ray'
```

**Comment by tgaddair** [comment]
I would say I agree. Right now we have a lot of extras in Ludwig, but we don't have adequate test coverage to know if those extra dependencies always life off the critical import path. One solution to this problem would be to add a run of the unit tests that installs Ludwig with minimal dependencies and a limited set of end-to-end tests to make sure everything works.

At the same time, I think text, image, and ray are all very core features a lot of people use and are deeply ingrained in the code, so making them extras puts a lot of additional work on our side to isolate them from the critical path. It may not be worth the effort if 90% of users are using them anyway.

**Comment by tgaddair** [comment]
Fixed by #1939.
